File: apps/demo-be-python-fastapi/tests/integration/steps/admin_steps.py
"""BDD step definitions for admin feature."""

from fastapi.testclient import TestClient
from pytest_bdd import given, parsers, scenarios, then, when

from tests.integration.conftest import GHERKIN_ROOT
from tests.integration.steps.security_steps import _ADMIN_PASSWORD, _register_and_promote_admin

scenarios(str(GHERKIN_ROOT / "admin" / "admin.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    parsers.parse('an admin user "{username}" is registered and logged in'),
    target_fixture="admin_tokens",
)
def admin_login(client: TestClient, username: str) -> dict:
    user_data = _register_and_promote_admin(client, username, _ADMIN_PASSWORD)
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": username, "password": _ADMIN_PASSWORD},
    )
    assert resp.status_code == 200
    return {**resp.json(), "admin_id": user_data["id"]}


@given(
    parsers.parse('users "{a}", "{b}", and "{c}" are registered'),
    target_fixture="registered_users",
)
def register_multiple_users(client: TestClient, a: str, b: str, c: str) -> list:
    users = []
    for username in [a, b, c]:
        resp = client.post(
            "/api/v1/auth/register",
            json={
                "username": username,
                "email": f"{username}@example.com",
                "password": _PASSWORD,
            },
        )
        assert resp.status_code == 201
        users.append(resp.json())
    return users


@given(
    parsers.parse('"{username}" has logged in and stored the access token'),
    target_fixture="alice_tokens",
)
def alice_login_for_admin(client: TestClient, username: str) -> dict:
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": username, "password": _PASSWORD},
    )
    assert resp.status_code == 200
    return resp.json()


@given("alice's account has been disabled by the admin")
def disable_alice_by_admin(client: TestClient, registered_users: list, admin_tokens: dict) -> None:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    resp = client.post(
        f"/api/v1/admin/users/{alice['id']}/disable",
        json={"reason": "Test disable"},
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )
    assert resp.status_code == 200


@given("alice's account has been disabled")
def alice_account_disabled(client: TestClient, registered_users: list, admin_tokens: dict) -> None:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    resp = client.post(
        f"/api/v1/admin/users/{alice['id']}/disable",
        json={"reason": "Initial disable"},
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )
    assert resp.status_code == 200


# --- When steps ---


@when("the admin sends GET /api/v1/admin/users", target_fixture="response")
def admin_list_users(client: TestClient, admin_tokens: dict):  # type: ignore[no-untyped-def]
    return client.get(
        "/api/v1/admin/users",
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )


@when(
    "the admin sends GET /api/v1/admin/users?email=alice@example.com",
    target_fixture="response",
)
def admin_search_users_by_email(client: TestClient, admin_tokens: dict):  # type: ignore[no-untyped-def]
    return client.get(
        "/api/v1/admin/users",
        params={"email": "alice@example.com"},
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )


@when(
    parsers.parse("the admin sends POST /api/v1/admin/users/{{alice_id}}/disable with body {body}"),
    target_fixture="response",
)
def admin_disable_alice(client: TestClient, registered_users: list, admin_tokens: dict, body: str):  # type: ignore[no-untyped-def]
    import json

    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    data = json.loads(body)
    return client.post(
        f"/api/v1/admin/users/{alice['id']}/disable",
        json=data,
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )


@when(
    "the client sends GET /api/v1/users/me with alice's access token",
    target_fixture="response",
)
def get_me_alice_token(client: TestClient, alice_tokens: dict):  # type: ignore[no-untyped-def]
    return client.get(
        "/api/v1/users/me",
        headers={"Authorization": f"Bearer {alice_tokens['access_token']}"},
    )


@when(
    parsers.parse("the admin sends POST /api/v1/admin/users/{{alice_id}}/enable"),
    target_fixture="response",
)
def admin_enable_alice(client: TestClient, registered_users: list, admin_tokens: dict):  # type: ignore[no-untyped-def]
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    return client.post(
        f"/api/v1/admin/users/{alice['id']}/enable",
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )


@when(
    parsers.parse("the admin sends POST /api/v1/admin/users/{{alice_id}}/force-password-reset"),
    target_fixture="response",
)
def admin_force_password_reset(client: TestClient, registered_users: list, admin_tokens: dict):  # type: ignore[no-untyped-def]
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    return client.post(
        f"/api/v1/admin/users/{alice['id']}/force-password-reset",
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )


# --- Then steps ---


@then(
    'the response body should contain at least one user with "email" equal to "alice@example.com"'
)  # noqa: E501
def check_alice_in_results(response) -> None:  # type: ignore[no-untyped-def]
    body = response.json()
    users = body.get("data", [])
    assert any(u.get("email") == "alice@example.com" for u in users), (
        f"alice@example.com not found in users: {users}"
    )


@then('alice\'s account status should be "disabled"')
def check_alice_disabled(client: TestClient, registered_users: list, admin_tokens: dict) -> None:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    resp = client.get(
        "/api/v1/admin/users",
        params={"email": "alice@example.com"},
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )
    body = resp.json()
    users = body.get("data", [])
    alice_info = next((u for u in users if u["username"] == "alice"), None)
    assert alice_info is not None
    assert alice_info["status"].upper() == "DISABLED"


@then('alice\'s account status should be "active"')
def check_alice_active(client: TestClient, registered_users: list, admin_tokens: dict) -> None:
    alice = next((u for u in registered_users if u["username"] == "alice"), None)
    assert alice is not None
    resp = client.get(
        "/api/v1/admin/users",
        params={"email": "alice@example.com"},
        headers={"Authorization": f"Bearer {admin_tokens['access_token']}"},
    )
    body = resp.json()
    users = body.get("data", [])
    alice_info = next((u for u in users if u["username"] == "alice"), None)
    assert alice_info is not None
    assert alice_info["status"].upper() == "ACTIVE"

File: apps/crud-be-ts-effect/tests/integration/steps/admin.steps.ts
import { Given, When, Then } from "@cucumber/cucumber";
import assert from "node:assert/strict";
import type { CustomWorld } from "../world.js";
import { promoteToAdmin } from "../hooks.js";

// ---- Background ----

Given("an admin user {string} is registered and logged in", async function (this: CustomWorld, username: string) {
  const email = `${username}@example.com`;
  const password = "Str0ng#Pass1";

  // Register
  const res = await this.post("/api/v1/auth/register", { username, email, password });
  if (res.status === 201) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    this.userIds.set(username, (res.body as any).id as string);
  } else if (res.status !== 409) {
    throw new Error(`Failed to register admin ${username}: ${JSON.stringify(res.body)}`);
  }

  // Promote to ADMIN via direct DB access
  await promoteToAdmin(username);

  // Login
  const loginRes = await this.post("/api/v1/auth/login", { username, password });
  if (loginRes.status !== 200) {
    throw new Error(`Failed to login admin ${username}: ${JSON.stringify(loginRes.body)}`);
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const tokens = loginRes.body as { accessToken: string; refreshToken: string };
  this.tokens.set(`${username}_access`, tokens.accessToken);
  this.tokens.set(`${username}_refresh`, tokens.refreshToken);
  this.context[`${username}_password`] = password;
});

Given(
  "users {string}, {string}, and {string} are registered",
  async function (this: CustomWorld, u1: string, u2: string, u3: string) {
    for (const username of [u1, u2, u3]) {
      const email = `${username}@example.com`;
      const password = "Str0ng#Pass1";
      const res = await this.post("/api/v1/auth/register", { username, email, password });
      if (res.status === 201) {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        this.userIds.set(username, (res.body as any).id as string);
      } else if (res.status === 409) {
        // Already registered — get ID via admin list
        const adminToken = this.tokens.get("superadmin_access") ?? "";
        const listRes = await this.get(`/api/v1/admin/users?search=${email}`, adminToken);
        if (listRes.status === 200) {
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          const data = (listRes.body as any).content as Array<{ id: string; username: string }>;
          const user = data.find((u) => u.username === username);
          if (user) {
            this.userIds.set(username, user.id);
          }
        }
      } else {
        throw new Error(`Failed to register ${username}: ${JSON.stringify(res.body)}`);
      }
      this.context[`${username}_password`] = password;
    }
  },
);

// ---- Account disabled ----

Given("alice's account has been disabled by the admin", async function (this: CustomWorld) {
  const adminToken = this.tokens.get("superadmin_access") ?? "";
  const aliceId = this.userIds.get("alice") ?? "";
  const res = await this.post(`/api/v1/admin/users/${aliceId}/disable`, { reason: "Test" }, adminToken);
  if (res.status !== 200) {
    throw new Error(`Failed to disable alice: ${JSON.stringify(res.body)}`);
  }
});

Given("alice's account has been disabled", async function (this: CustomWorld) {
  const adminToken = this.tokens.get("superadmin_access") ?? "";
  const aliceId = this.userIds.get("alice") ?? "";
  const res = await this.post(`/api/v1/admin/users/${aliceId}/disable`, { reason: "Test" }, adminToken);
  if (res.status !== 200) {
    throw new Error(`Failed to disable alice: ${JSON.stringify(res.body)}`);
  }
});

// ---- When: admin requests ----

// Generic: the admin sends GET <path>
When(/^the admin sends GET (.+)$/, async function (this: CustomWorld, path: string) {
  const adminToken = this.tokens.get("superadmin_access") ?? "";
  this.response = await this.get(path, adminToken);
});

// Disable alice with reason
When(
  /^the admin sends POST \/api\/v1\/admin\/users\/\{alice_id\}\/disable with body \{ "reason": "([^"]*)" \}$/,
  async function (this: CustomWorld, reason: string) {
    const adminToken = this.tokens.get("superadmin_access") ?? "";
    const aliceId = this.userIds.get("alice") ?? "";
    this.response = await this.post(`/api/v1/admin/users/${aliceId}/disable`, { reason }, adminToken);
  },
);

// Enable alice
When(/^the admin sends POST \/api\/v1\/admin\/users\/\{alice_id\}\/enable$/, async function (this: CustomWorld) {
  const adminToken = this.tokens.get("superadmin_access") ?? "";
  const aliceId = this.userIds.get("alice") ?? "";
  this.response = await this.post(`/api/v1/admin/users/${aliceId}/enable`, {}, adminToken);
});

// Force password reset for alice
When(
  /^the admin sends POST \/api\/v1\/admin\/users\/\{alice_id\}\/force-password-reset$/,
  async function (this: CustomWorld) {
    const adminToken = this.tokens.get("superadmin_access") ?? "";
    const aliceId = this.userIds.get("alice") ?? "";
    this.response = await this.post(`/api/v1/admin/users/${aliceId}/force-password-reset`, {}, adminToken);
  },
);

// Unlock alice
When(/^the admin sends POST \/api\/v1\/admin\/users\/\{alice_id\}\/unlock$/, async function (this: CustomWorld) {
  const adminToken = this.tokens.get("superadmin_access") ?? "";
  const aliceId = this.userIds.get("alice") ?? "";
  this.response = await this.post(`/api/v1/admin/users/${aliceId}/unlock`, {}, adminToken);
});

// ---- Then: admin assertions ----

Then(
  /^the response body should contain at least one user with "([^"]*)" equal to "([^"]*)"$/,
  function (this: CustomWorld, field: string, value: string) {
    assert.ok(this.response !== null);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const data = (this.response?.body as any)?.content as Array<Record<string, string>>;
    assert.ok(Array.isArray(data));
    const found = data.some((item) => String(item[field]) === value);
    assert.ok(found);
  },
);

// Note: "alice's account status should be {string}" is defined in security.steps.ts
// Note: "the response body should contain {string} equal to {string}" is defined in auth.steps.ts
